### ISSUE TITLE
Off-by-one in counsel-yank-pop selection

### ISSUE THREAD

**Issue by ambihelical** [abo-abo/swiper#1371]
If I understand the recent changes correctly, the current kill-ring-yank-pointer position should be preselected. This seems to always be one off.  If I have three lines which I've killed in order:

one
two
three

After killing "one", the counsel-yank-pop shows  "one" and it is preselected.
After killing "two":
```
    two <- should be selected
    one <- selected
```
After killing "three":
```
    three  <- should be selected
    two  <- selected
    one

```
When I use the "rotate" command to change the current kill-ring-yank-pointer, it never sets the selection to the item selected, but is always one off.   If I select "three", and use rotate (M-o r) to set the pointer, the next counsel-yank-pop shows "two" selected.  If I select "two" and rotate, the next counsel-yank-pop shows "one" selected.  If I select "one", the next invocation shows "three" selected. 

For this issue I was using commit 5c096a2.

**Comment by abo-abo** [comment]
@basil-conto Can you have a look? Off-by-one results from the change in `interactive` spec - `arg` is usually 1 when it's expected to be zero.

**Comment by basil-conto** [comment]
#### TL;DR

Preselecting the next-to-last kill is intended.

#### Rationale

The recent changes to `counsel-yank-pop` try to mirror the default `yank`, `yank-pop`, `rotate-yank-pointer`, etc. functionality as close as possible, whilst providing convenient realtime visualisation/completion and custom actions via the Ivy/Counsel interface.

As [`(emacs) Yanking`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Yanking.html) explains in further detail, the general idea is that `yank` retrieves the last kill, whereas `yank-pop` is for accessing [earlier kills](https://www.gnu.org/software/emacs/manual/html_node/emacs/Earlier-Kills.html). This is also summarised in the docstring of `yank-pop`:

    yank-pop is an interactive compiled Lisp function in ‘simple.el’.

    It is bound to M-y.

    (yank-pop &optional ARG)

    Replace just-yanked stretch of killed text with a different stretch.
    This command is allowed only immediately after a ‘yank’ or a ‘yank-pop’.
    At such a time, the region contains a stretch of reinserted
    previously-killed text.  ‘yank-pop’ deletes that text and inserts in its
    place a different stretch of killed text.

    With no argument, the previous kill is inserted.
    With argument N, insert the Nth previous kill.
    If N is negative, this is a more recent kill.

    The sequence of kills wraps around, so that after the oldest one
    comes the newest one.

    This command honors the ‘yank-handled-properties’ and
    ‘yank-excluded-properties’ variables, and the ‘yank-handler’ text
    property, in the way that ‘yank’ does.

The optional prefix argument defaults to 1, meaning yank the next-to-last kill. If you want to yank the last kill, you should pass it a prefix argument of zero.

The current implementation of `counsel-yank-pop` mirrors this by preselecting what vanilla `yank-pop` would have yanked. So it doesn't "select" the "wrong" candidate, it "preselects" the "default" candidate.

**Comment by abo-abo** [comment]
@basil-conto Thanks for the explanation. It makes sense in a way.

Yet in another way it doesn't. Using <kbd>M-y RET</kbd> instead of <kbd>C-y</kbd> seems like a nice use case. Of course, if you did <kbd>M-w</kbd> seconds ago, you know what's on the top of the kill ring and you can do <kbd>C-y</kbd> directly. But if you're just fairly sure that the thing you want is on the top of the kill ring, <kbd>M-y RET</kbd> gives you a preview, an option to back out with <kbd>C-g</kbd>, and an option to select something else.

Currently, the above use case was moved to the <kbd>M-y C-p RET</kbd> binding. I'm not sure it's worth the consistency with the defaults.

**Comment by basil-conto** [comment]
> It makes sense in a way. Yet in another way it doesn't.

If you'll pardon the facetious retort, it doesn't make sense to discuss whether this makes sense. :)

Here's my understanding of the situation. Please to confirm/comment.

* Emacs provides intricate, central and well-documented idioms and interfaces around killing and yanking, with defaults and conventions which support a particular workflow.
* Counsel can extend the user-friendliness and flexibility of many such systems via Ivy filtering, actions, etc.
* The command `counsel-yank-pop` previously did no preselecting, meaning the first completion candidate was preselected by default.
* The command's recent reimplementation was intended to reuse the existing Emacs system but make it more Counsel-y. Part of using the existing system is respecting the `yank-pop` idiom of doing what's in the name and *popping* the `kill-ring`. This was not just to satisfy the author's compulsive pedantry, but also to satisfy a user request (#1190).
* Now, users (and owner :) seem (because no-one has stated it clearly) to be requesting that Counsel should return to and/or also support preselecting the last kill.

So my conclusion is simply do the Counsel-y thing and add another user toggle which defaults to the old, non-Emacs behaviour, right?

**Comment by dieggsy** [comment]
@basil-conto A user toggle/custom var seems like a good compromise here

**Comment by abo-abo** [comment]
@basil-conto You're right. `counsel-yank-pop` does say

>   "Ivy replacement for `yank-pop'.

Could you add a defcustom for this, with the current behavior as the default?

**Comment by basil-conto** [comment]
@abo-abo 

> Could you add a defcustom for this

I can't promise I can, but I can promise to try. :)

> with the current behavior as the default

Are you sure you don't prefer the old behaviour as default?

**Comment by abo-abo** [comment]
> Are you sure you don't prefer the old behaviour as default?

I was planning to customize it in my config, in the name of `counsel-yank-pop` staying true to its name.

We can change it later if needed.

**Comment by articuluxe** [comment]
@basil-conto I just wanted to say, thanks for this change.  It makes this feature usable for me and it is a big improvement IMO to make the functionality consistent with standard emacs usage.

**Comment by basil-conto** [comment]
@articuluxe Glad it's useful for you; thanks are due to others, though. :)
